### PR TITLE
Type useSWR fetcher

### PR DESCRIPTION
**Pull request by promer94**
- [x] improve the type check process of the project.  Currently we only do type check in src and test folder.
- [x] support ```useSWR({a: number, b: string}, (args) => args)```
- [x] support ```useSWR(cond() ? {a: number, b: string} : null , (args) => args)```
- [x] fix current build error 
- [x] find a batter way to test type

## Playground Link
[useSWR](https://www.typescriptlang.org/play?#code/MYewdgzgLgBArhApgZQOoCUBcM3oBIggDWMAvDABQCUZAfDAN4BQMrMUAFgE4gDuMYRPwCiXHlwoByMCFgBLALYAHADYLEYKIgAmkqkwC+TJqEixT27NTowARoRWIAhmDKUapes1ace-QSJiIBLSsjCKquqaOnqGxgBmcGDAUHLg8EgAKojQ1t4wLBkoGBQARAD0Tkpy5QiIXKUANJRwXCoe9K3tANyFdbgUFtYA-DAVVTV1DTDYYHAqKs0UXR3wbVS9fUgDDDBO2KUAjE122LvABwDMpTAGt0srNiu9rP0lQzSju-tjx822ZxgFzG12alhgACZmrc7rN5osYEREABPGxI5EbYyvbYlADa3wOf1OjCBVxudwMzVxhwhl0OzUkAA5mZIALrNAkwGmXW7s1gUAB0QqcXAA5hBVrihQKReLWZjscV0INwNoRjB8XtCScASTgaVrrd7hrufSYEyWXzOdzeTMBPDGoLhWKJTYpc65ZisUUBpqfkcdYD9YaKVTTQzmYy2RytVzabanBAgeBoEtpbLXZ4NemXfKXj73qr1X7tf8g2SjZSTbSzRao1bYzaDKy9knTNAZnMFmmPZn6O6ZbmvYqBtYs+NqrUkA1HY8s88tkqKGP6B8YKMJ5Npzc4d3lusnutNiOSivKJyA2W9RWKVRoei0SiFQXlWe16MKBeibrzjfbjRdwRB8s3RYcX2XSUv0Da8QXJY1qRrCNLRjH4m3ZYD6FAxdR1Wd8NSgq9f1gysw0Q81I2jElUPjZs7S7ICUUfDFvTeV9INjS9iSIg04KrBC6SQ+sUOwNDW2TMxGgwxEn2w09cKLT58I479y2I0NqwE8jkKokSaJbRNxI7QDmikrCjCYRAAA8lGCWAoGRJREBgdAcnmKAAB5MjcJIiBkXgwHocgvIAHxgAAFHgFDkJBPNoCzrNs9gHKczI4FURAAGlGPIXEXGRZppR8vywFxVkW1CrhnG0cAVFRXKwHymBCrAXy+BKsr4psrg7OSmAADUnBUOBMuymBoC4OQwFFGBQvomaYFS9KstRUKXNALhtHcvLmjyuKrK6nrHJgZa3AGoaRpWygzzO4blv0TrEvso6ADFECgYAOHqdyABEnCgJxvJa4rmgAQRdGArK0MBtCTE7yGWwLCnKAAqZHChgZH3DdXZ4kIbBxsm6aq12WwRThBRbHqW1ZvhdG0bYTGzwoJTcZAfGoAmqbjRJsn7Qpqm7n0tsU1gGmFhgfQGfKQowfFCHLKhmGrtWSqnGqsBauzIVJviKmMvK+0FkltgPydQdxWwAd9dWFyIDcn6-qcWhjdYbAUfptgMZZvGxo5wnubsXm5n5rgjQN+j0dYD2Ge9tnfc5omY1Jrhycp0PBbE9tRcNlRI4x6XPdlpNIY0JWKFV9XNYHHW9flSPTZzC2tYFa2bFt+3fv+53I7d1G88Z1YCa5sXc89r2zzW4JNu2vYGvoEf+4Ltgi-lxWkwgmwa9DkKc5d9dKEbiBLcyeU29clQPM7p2e5gd3+9YIfppH+bJ42raGsaXbF8jleS+hpMI6e1GIIAAbvUG+v8Fal3XlvBae8PyH2PqfLM7cL4Oy7jfUB4DjAPW6uEaIXB4hOGAE5XABBiCMEKOggG5AZ6QLXsdUaCMKAZmwEXKg2Ar5UKvm4Oh4M-5Kzhow5EtAKCR1YTAIujRI7xDANgV671PpcGoaDF088c6FA4TALhrBqG8I-pI-hUD-7CLcMwyOiCNRFwNriKRMAFEfS+lfVR4p1H0VZJozhjs4g4KYEAA)
[useSWRInfinite](https://www.typescriptlang.org/play?#code/KYDwDg9gTgLgBDAnmYcAqBXMAbYBpYROAXjgG0BDAO0QBo4A6JjKgayogHcqyBdXuAB84UYBQAmEKtiKUa9JgxbsuPfgChQkWAmSoAahWwZ8hEnADOMKAEsqAcyFwqGbNieYcposIBKwAGNocQAeajo4cIA+dU1waHgkFDgCRAAZCAlgKBCAQSh7CzhQGGAqcSLDY29zKpNUqJJ1OCcACla7cVAALmcMAFsAI2z6MFEANxsIDAsABQp7YAARChgKXvCnFzcAShJG-MKd5q3XbFik1H8LVxgQtHNlDm5G0gfhWagIfpsLYHuYnFtIk9HAAJJUABmdhspQAYsAYAEABbZEInQ5FEplCpwOo1Uj41K0E4rNbmcLqV5wTHFEClcpFVqiCRSGTkRR2SHZFK8Y4tAD8cHaigoBQsvTIijwfP2cGutxCZIoUX5cF6tOxjL6bhOQqowHG2ROGvFdIZuI6UJ5aDVQtaovFkrQsuIjQV2DuypiLV6BqNUFidlKUEhFACqBmwAAygB1XwQ6FUWGoADeJyVqwo9E19JxlSM9UIUVaYsKvVSGSyOUxUXokKovUTMPhiJRaMx9G9p12vWVfHUAF9YkEqFY4FG4wmoS3gL1J-Hm8nSuZWns3XB0y0YMivpxnMB9wBRKBfKCtADkHHgNn6OH6ZVK4gvx2Ho-Ho-EvTXcsGEAguDUKu66NFu267lwB7Hqe0CXtecC3vej7AM+exDuokIsAEMBTFQE5-GgwBWD+YEtAu05Jim7SdKAowTFMMzzIsyogXAF4APQUGANjsVGUAXvQpasRQapkX8U5LlRVpdCAdGGgxcwLMsWasZ+P5ChxXE8XxF5+mcgkiXKIknGJMaLjOy7ANR5S0XAYzydMinMSpcqtKmcCQv+vQXgAjAATAAzAJcCDGKvQACwAKwAGxwIOOwGaxbnxZERSpp5EC9FYtgOHFcCifhZkUbO1kyXJkyOUxylrKpUjiOpwruRl3n+UF9ChVAEUxXFex6W4iVGQV5GSaUpW2fZFWMUpLFymQTVeWxrXBR1XWxYO9C+YFwVRdF9BkJtQX8IJrCEBYrFzTYX6LVt9BUBQD4tYFfk+UFg58iZhUSRZUk0bJdn0ZV00uRual7AKc0eQtB3LWFcA7XFG1LfQO17dD-Dqjq2AGbQgy0AEtDiOd2a4-j4jvS0plfZRo3SeNANTc5NWzfNmXXW1IWw-D61s9tMWo0t6MUEU77wNjJME0TON4wT5MU8N300795UKVVM0bnAoNwEKEPNTz7Wc913PQ8jfPkGjAhCxrUjjn1WPCsT0uE7NDuk+9g5AA)

### CODE REVIEW

File: src/types.ts [hunk 1]
```diff
@@ -1,6 +1,23 @@
 import * as revalidateEvents from './constants/revalidate-events'
 
-export type Fetcher<Data> = (...args: any) => Data | Promise<Data>
+export type Result<T = unknown> = T | Promise<T>
+
+export type Fetcher<
+  Data = unknown,
+  Args extends Key = Key
+> = Args extends (() => [infer R, ...infer K] | readonly [infer R, ...infer K])
```
Comment (shuding) [inline src/types.ts:8]: Why do we need `infer R, ...infer K` here instead of just one `...infer`?

**Review by shuding — COMMENTED** [review]
This is amazing work, appreciate it! ❤️

File: src/types.ts (hunk 1 above)
Comment (promer94) [reply src/types.ts:8]:
```typescript
useSWR(key: string | null)
useSWR(key: Record<any, any>)
useSWR(...keys: [any, any[]])
// typescript cant get the first and second inferred correctly by [...infer K]
// so i use a [any, unknown[]] here as a workaround here
// i cant give you a clear explanation for this because i don't know much about this part 😂
```

File: src/use-swr.ts [hunk 2]
```diff
@@ -175,7 +175,8 @@ export const useSWRHandler = <Data = any, Error = any>(
 
           // Start the request and keep the timestamp.
           CONCURRENT_PROMISES_TS[key] = startAt = getTimestamp()
-          newData = await (CONCURRENT_PROMISES[key] = fn.apply(fn, fnArgs))
+          // @ts-ignore FIXME LATER
```
Comment (huozhi) [inline src/use-swr.ts:178]: the dependency of typescript eslint `tsutils` didn't resolve the right version of tslib which lead to the error.
you can add a resolution for tslib to 2.3.0 in root top package.json. but might be hacky.

I submit a PR to it but didn't get merged...

Comment (huozhi) [reply src/use-swr.ts:178]: I separate some upgration changes to https://github.com/vercel/swr/pull/1486

**Comment by promer94** [comment]
I tried to make useSWRInfinite also work but i can't find a perfect implementation now.

```typescript
useSWRInfinite((index, previousPageData) => [{ foo: '123', bar: 456 }, '123', 456, ['123']], (...keys) => [{id: '123', name: '123213'}])

// The keys and result of hooks will be correctly typed 
// but previousPageData would become any 
```

File: package.json [hunk 3]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "swr",
-  "version": "1.1.0-beta.0",
+  "version": "1.1.0-beta.1",
```
[the remark this answers is not in the document]
Comment (promer94) [reply package.json]: fixed

File: package.json [hunk 4]
```diff
@@ -46,7 +46,11 @@
     "build:immutable": "bunchee index.ts --cwd immutable -m --no-sourcemap",
     "prepublishOnly": "yarn clean && yarn build",
     "publish-beta": "yarn publish --tag beta",
-    "types:check": "tsc --noEmit",
+    "types:check": "yarn types:check:core && yarn types:check:infinite && yarn types:check:immutable && yarn types:check:fetcher",
+    "types:check:core": "tsc --noEmit",
+    "types:check:infinite": "cd infinite && tsc --noEmit",
+    "types:check:immutable": "cd immutable && tsc --noEmit",
+    "types:check:fetcher": "cd test-type && tsc --noEmit",
```
Comment (huozhi) [inline package.json:53]: we can reduce these scripts by running type checking on top root dir to include all assets. I filed https://github.com/vercel/swr/pull/1492 to reduce them

```suggestion
    "types:check": "tsc --noEmit",
```

File: test-type/fetcher.ts [hunk 5]
```diff
@@ -0,0 +1,164 @@
+import useSWR from 'swr'
+type ExpectType = <T>(value: T) => void
+const expectType: ExpectType = () => {}
+
+const cond: () => boolean = () => true
```
Comment (huozhi) [inline test-type/fetcher.ts:5]:
```suggestion
const truthy: () => boolean = () => true
```

File: src/utils/with-middleware.ts [hunk 6]
```diff
@@ -9,10 +9,10 @@ export const withMiddleware = (
 ): SWRHook => {
   return <Data = any, Error = any>(
     ...args:
-      | readonly [Key]
-      | readonly [Key, Fetcher<Data> | null]
-      | readonly [Key, SWRConfiguration | undefined]
-      | readonly [Key, Fetcher<Data> | null, SWRConfiguration | undefined]
```
Comment (huozhi) [inline src/utils/with-middleware.ts:15]: why removing `readonly`?

File: src/use-swr.ts [hunk 7]
```diff
@@ -173,7 +173,7 @@ export const useSWRHandler = <Data = any, Error = any>(
 
           // Start the request and keep the timestamp.
           CONCURRENT_PROMISES_TS[key] = getTimestamp()
-          CONCURRENT_PROMISES[key] = fn.apply(fn, fnArgs)
+          CONCURRENT_PROMISES[key] = fn(...fnArgs)
```
Comment (huozhi) [inline src/use-swr.ts:176]: I think we want to keep apply to avoid tslib applying object spread transform onto it

File: src/utils/with-middleware.ts (hunk 6 above)
Comment (promer94) [reply src/utils/with-middleware.ts:15]: This was used to make tuple inference works before TS 4.0. 
https://github.com/vercel/swr/pull/741/files#diff-34add1ba5917e182ff385ebefbcff5dd9b012723f395e89d77548c26c9f8735cR48-R52

Since we have upgraded the typescript version we don't need it any more

File: src/use-swr.ts (hunk 7 above)
Comment (promer94) [reply src/use-swr.ts:176]: should i also change here ?
https://github.com/vercel/swr/blob/86fb5da797a15079172cb079e367586ef7cb4bdd/infinite/index.ts#L138

File: package.json (hunk 4 above)
Comment (promer94) [reply package.json:53]: But it would create test-type/fetcher.d.ts in dist folder. Could it  be avoided ?

File: src/use-swr.ts (hunk 7 above)
Comment (huozhi) [reply src/use-swr.ts:176]: Yeah I think that one is also safe to change to `apply` call

**Comment by promer94** [comment]
## limitations
* user has to type previousPageData explicitly
```typescript
// index key and result the hook will be inferred correctly 
useSWRInfinite((index, previousPageData: {id: number, name: string }) => index, (key) => [{id: 123, name: '123213'}])
```
* key will not be inferred if user type hook's data using generic
```typescript
// key would become any 
useSWRInfinite<{id: number, name: string}[]>((index) => index, (key) => [{id: 123, name: '123213'}])
useSWR<{id: number, name: string}>('/api/', (key) => ({id: 123, name: '123213'}))
```

**Comment by shuding** [comment]
I think those limitations are good for `useSWRInfinite`, as it has deeper type issues and we will leave it for the future. But why is "key will not be inferred if user type hook's data using generic" for `useSWR`? I think the key should still be able to be inferred theoretically.

**Comment by promer94** [comment]
I have updated the example. 
I mention this because sometimes i saw people write ```useSWR<{foo: string}>(key, fn)``` for convenience. In this way, they don't need to type their fetcher.

File: src/index.ts [hunk 8]
```diff
@@ -20,5 +20,7 @@ export {
   SWRHook,
   Fetcher,
   MutatorCallback,
-  Middleware
+  Middleware,
+  ValueKey,
+  Result
```
Comment (shuding) [inline src/index.ts:25]: Are those 2 types necessary for the users? Or is it just for `useSWRInfinite`?

File: src/types.ts [hunk 9]
```diff
@@ -1,6 +1,46 @@
...
+       *
+       * the fetcher would be
+       * (arg0: string) => any | (arg0: { foo: string, bar: number }) => any
+       * so we add this condition to make (arg0: string) => any to be never
+       */
+    ? Args extends (Record<any, any> | null)
+      ? never
+      : (...args: [string]) => Result<Data>
+    : Args extends (infer T)
+    ? (...args: [T]) => Result<Data>
```
Comment (shuding) [inline src/types.ts:42]: Can those 2 cases (string, record) be merged as one?

```ts
Args extends (infer T)
  ? (...args: [T]) => Result<Data>
  : never
```

**Review by shuding — COMMENTED** [review]
Other than these 2, all looking great!

File: src/index.ts (hunk 8 above)
Comment (promer94) [reply src/index.ts:25]: Currently it is just for ```useSWRInfinite ```. 
But it might be useful for user to build middleware like ```useSWRMutation```

File: src/types.ts (hunk 9 above)
Comment (promer94) [reply src/types.ts:42]: The simple answer is no.

But i just realize it could have a batter implementation. 
## TLDR

```typescript
// When conditional types act on a generic type, they become distributive when given a union type
// So if we have type like.
type Fn<Args extends (string | null | Record<any,any>) = string | null | Record<any,any>> = 
  Args extends (infer T)
    ? (...args: [T]) => any
    : never

// the result here is actully a union type 
// (string extends (infer T) ? (...args: [T] =>  any) : never) | (null extend (infer T) ? (...args: [T] =>  any) : never)  
// so the result is (args: string) => any | (args: null) => any
type Test = Fn<string | null>


// we have to give it more specific condition
type Fn2<Args extends (string | null | Record<any,any>) = string | null | Record<any,any>> = 
  Args extends (string | null)
    ? (...args: [string]) => any
    : Args extends (infer T)
    ? (...args: [T]) => any
    : never

// Now These would work
type Test2 = Fn2<string>
type Test3 = Fn2<string | null>
type Test4 = Fn<Record<string, number>>

// (args: null) => any | (args: Record<string, number>) => any
type Test5 = Fn<Record<string, number> | null>


// add condition to remove (args: null) => any
type Fn3<Args extends (string | null | Record<any,any>) = string | null | Record<any,any>> = Args extends null
      ? never
      : Args extends (infer T)
    ? (...args: [T]) => any
    : never

type Test6 = Fn3<string>
type Test7 = Fn3<string | null>
type Test8 = Fn3<Record<string, number>>
type Test9 = Fn3<Record<string, number> | null>
```

[Playground](https://www.typescriptlang.org/play?#code/PTAEHUAsFMDtQMYHtYBMCWAXdKCGAbUTATwAdoBnUXBTUFa0AczmgCd0EizoAaImMVAAjaMgC20UBgqYOwgK7YAblIDuMeE3Sr4uUAtg54JcgCgQoAMpJQ6AGag1UyLlXdyofOgDW0AHRmplIAYrAAPACCbExU0AAemHCoVAAUshywTKAAPqCwCviEeQBKYkhsqOG4sMS8NcQAfACUoAC8oBnoWbn5hcWgZciV1bX1tY2N7aBmoKDRsaAJSWhp3fbsoAAqzbNzoAD8oKn+p7gxFABcoADaWwC6rW1TDXtz17DQqmxmFmCYMFAbEohToMGBdioNEw-SE+kMxg8Uj+xy6PWWyTWsA2bG2rSOJzOF2ud3u7Sm1FqrQ+X3YrTyqQKRSWiWSx3Wmx2h2Op3851iJIe5LmDWp+VpbFaM0sFFsAKkwIooMhx35V06cm6TCeL1qvVSao+-R1lOIQR420odA6YXCaOyeSZ+EavxRzlArncmFs2ncWFA4gqUgo5AQDk4iBQGGwKHNnjCACYohcWSsUqjNT1Hf1ekMKlUGuMmk8NZkHX1maVyiNCw1JtM9gs4qzVhmy70nbt9tzCXzibd7Y9ha9u9cm6nMezsZyu-sCbzDbcHiaR-sad9fpYAHJINTbGAUdRIQqoJwVHxxqRbK0J6aJu2Zpgu4KW2QAZjvsCT9o7-WfFuvWQABZP3CPMRntfgCnEUQ2EmTcwANftO2HPUGUXcCqkgvoYPYFpULNF9AMwABWUDMIfMsoIUXC4N-IoXQQ6hUFPZA0CwRFvSBaBA3cJCBQrfAV1qS9QDCN9k0WDFW3SR96IGCjawmEsf2zStBmrAsxjrKYOnHaT0ydN453FDdu3eeYUwMrEcTxYz5yJATSWEs1R1M9hfiIq0ADZPwk+1-08YiAHY-MorV5MCq8rQADjCijsOg2D4K82QAE54s08Ksmo2ipjU50gA)
[Distributive Conditional Types](https://www.typescriptlang.org/docs/handbook/2/conditional-types.html#distributive-conditional-types)

File: src/index.ts (hunk 8 above)
Comment (shuding) [reply src/index.ts:25]: Makes sense, thanks!

File: src/types.ts (hunk 9 above)
Comment (shuding) [reply src/types.ts]: Great idea! 👍

**Review by shuding — APPROVED** [review]
🚀

**Comment by huozhi** [comment]
Thank you! This is awesome 🚀

**Comment by Kamahl19** [comment]
I am unable to use generics after this change

```
  const { data } = useSWR<boolean, ErrorResponse>(
    accountName.length >= ACCOUNT_NAME_MIN_LENGTH
      ? [`/account/name-availability/${accountName}`, null]
      : null,
    fetcher
  );
```

I get 

```
Type '(url: string, token: string | null) => Promise<any>' is not assignable to type '((args_0: string) => FetcherResponse<boolean>) | ((args_0: any, ...args_1: unknown[]) => FetcherResponse<boolean>) | ((args_0: Record<any, any>) => FetcherResponse<...>) | ((args_0: string | ... 2 more ... | Record<...>) => FetcherResponse<...>) | null'
```

**Comment by huozhi** [comment]
@Kamahl19 Can you file another issue with the repro in codesandbox? Thanks

**Comment by Kamahl19** [comment]
https://codesandbox.io/s/great-andras-0m5bc?file=/src/index.tsx 

fetcher in config says `Type '(org: string, repo: string) => Promise<any>' is not assignable to type '(args_0: string) => any'.ts(2322)`
